### PR TITLE
Cover Dashboard page with tests

### PR DESCRIPTION
**Pull request by MBovtriukProvectus**
add test for ClustersWidget

### CODE REVIEW

**Review by GneyHabub — COMMENTED** [review]
There are some uncovered lines that can be easily covered: https://sonarcloud.io/code?id=provectus_kafka-ui_frontend&selected=provectus_kafka-ui_frontend%3Asrc%2Fcomponents%2FDashboard%2FClustersWidget%2FClustersWidget.tsx
Also, some user interaction can be simulated

**Comment by workshur** [comment]
Please add specs for Container and Dashboard component

**Comment by MBovtriukProvectus** [comment]
> Please add specs for Container and Dashboard component

done, if I correctly understand

**Review by GneyHabub — COMMENTED** [review]
What I meant by _simulating user input_ is that you can simulate click on the checkbox. This way you'll achieve higher coverage. See already existing tests to see how it can be done if you don't know.
Also, ever wondered why the function for testing is called `it`? That's because it is supposed to form a proper English sentence together with the first argument, which describes the matter of assessment. For example: `it('calls the fetch function', () => {...})`.
In addition, the Dashboard component remains untested. You are supposed to create a new spec file and test the Dashboard.tsx explicitly.

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  });
+
+  it('ClusterWidget can be truthy', () => {
+    const wrapper = component().find('ClusterWidget');
+    expect(wrapper).toBeTruthy();
+  });
+
+  it('columns can be truthy', () => {
+    const wrapper = component().find('.columns');
+    expect(wrapper).toBeTruthy();
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:29]:
```suggestion
    expect(component().exists('.columns')).toBeTruthy();
```

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+describe('ClustersWidget', () => {
+  it('render list length can be 2', () => {
+    const wrapper = component().find('ClusterWidget');
+    expect(wrapper.length).toBe(2);
+  });
+
+  it('ClusterWidget can be truthy', () => {
+    const wrapper = component().find('ClusterWidget');
+    expect(wrapper).toBeTruthy();
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:24]:
```suggestion
    expect(component().exists('ClusterWidget')).toBeTruthy();
    
```

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    <ClustersWidget
+      clusters={clusters}
+      onlineClusters={[onlineCluster]}
+      offlineClusters={[offlineCluster]}
+    />
+  );
+
+describe('ClustersWidget', () => {
+  it('render list length can be 2', () => {
+    const wrapper = component().find('ClusterWidget');
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:18]: it is not a wrapper...

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    expect(wrapper).toBeTruthy();
+  });
+
+  it('columns can be truthy', () => {
+    const wrapper = component().find('.columns');
+    expect(wrapper).toBeTruthy();
+  });
+
+  it('after clicking into input list length can be 1', () => {
+    const wrapper = component().find('input');
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:33]: const input

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidgetContainer.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    );
+    it(' is empty when no online clusters', () => {
+      const wrapper = mount(setupEmptyWrapper());
+      expect(wrapper.find('.is-success').text()).toBe('0');
+    });
+    it(' is empty when no offline clusters', () => {
+      const wrapper = mount(setupEmptyWrapper());
+      expect(wrapper.find('.is-danger').text()).toBe('0');
+    });
+  });
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidgetContainer.spec.tsx:21]: there is no need to test <ClustersWidget> here again

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import Dashboard from 'components/Dashboard/Dashboard';
+
+const component = shallow(<Dashboard />);
+
+describe('Dashboard', () => {
+  it('section can be truthy', () => {
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:8]: "section can be truthy" what does it mean?

I would suggest to use something like `it renders section`

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    const wrapper = component.find('.section');
+    expect(wrapper).toBeTruthy();
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:10]: pls use `.exists()` instead

**Review by workshur — CHANGES_REQUESTED** [review]

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    />
+  );
+
+describe('ClustersWidget', () => {
+  it('render list length can be 2', () => {
+    const wrapper = component().find('ClusterWidget');
+    expect(wrapper.length).toBe(2);
+  });
+
+  it('ClusterWidget can be truthy', () => {
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:22]: meaningless description

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  shallow(
+    <ClustersWidget
+      clusters={clusters}
+      onlineClusters={[onlineCluster]}
+      offlineClusters={[offlineCluster]}
+    />
+  );
+
+describe('ClustersWidget', () => {
+  it('render list length can be 2', () => {
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:17]: meaningless description

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    const wrapper = component().find('ClusterWidget');
+    expect(wrapper.length).toBe(2);
+  });
+
+  it('ClusterWidget can be truthy', () => {
+    const wrapper = component().find('ClusterWidget');
+    expect(wrapper).toBeTruthy();
+  });
+
+  it('columns can be truthy', () => {
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:27]: meaningless description

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    const wrapper = component().find('.columns');
+    expect(wrapper).toBeTruthy();
+  });
+
+  it('after clicking into input list length can be 1', () => {
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:32]:
```suggestion
 it('hides online cluster widgets')
```

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  });
+
+  it('ClustersWidgetContainer can be truthy', () => {
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:13]: meaningless description

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+    const wrapper = component.find('ClustersWidgetContainer');
+    expect(wrapper).toBeTruthy();
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:15]: pls use .exists()

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 1 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:29]: done

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 2 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:24]: done

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 3 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:18]: done

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 4 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:33]: done

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidgetContainer.spec.tsx (hunk 5 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidgetContainer.spec.tsx:21]: deleted

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx (hunk 7 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:10]: done

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx (hunk 13 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:15]: done

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx (hunk 6 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:8]: done

File: kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx (hunk 12 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/__test__/Dashboard.spec.tsx:13]: changed to "renders ClustersWidget"

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 11 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:32]: done

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 10 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:27]: changed to "renders columns"

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 9 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:17]: changed to "renders clusterWidget list"

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 8 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:22]: changed to "renders ClusterWidget"

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  );
+
+describe('ClustersWidget', () => {
+  it('renders clusterWidget list', () => {
+    const clusterWidget = component().find('ClusterWidget');
+    expect(clusterWidget.length).toBe(2);
+  });
+
+  it('renders ClusterWidget', () => {
+    expect(component().find('ClusterWidget')).toBeTruthy();
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:23]: .exists()

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  });
+
+  it('renders columns', () => {
+    expect(component().find('.columns')).toBeTruthy();
```
Comment (workshur) [inline kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx:27]: .exists()

Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx]: done

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx (hunk 14 above)
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx]: done

**Review by workshur — APPROVED** [review]

File: kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  });
+
+  it('renders columns', () => {
+    expect(component().exists('.columns')).toBeTruthy();
+  });
+
+  it('hides online cluster widgets', () => {
+    const input = component().find('input');
+    input.simulate('click');
+    expect(input.length).toBe(1);
```
[the remark this answers is not in the document]
Comment (MBovtriukProvectus) [reply kafka-ui-react-app/src/components/Dashboard/ClustersWidget/__test__/ClustersWidget.spec.tsx]: done

**Pull request merged**